Check for existing venv and Python versions

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Some systems may already have a Python venv in their core checkout. Even if not, the `python3` command on some systems may not be supported. (E.g., on Ubuntu 18.04.4, that command is version 3.6.9.)

Therefore, before attempting to create a new Python venv, first see if one exists. If so check its version (first activating it if necessary.) If it's ok, just use it as is. If one does not exist, then before using `python3`, check its version. If it's ok, use it. If not, then try `python3.8` or `python3.7`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pnbruckner committed:
Check for existing venv and Python versions

Before attempting to create a new Python venv, first see if one exists. If so check its version (first activating it if necessary.) If it's ok, just use it as is. If one does not exist, then before using `python3`, check its version. If it's ok, use it. If not, then try `python3.8` or `python3.7`.

frenck commented:
IMHO we should not try to handle venv in our scripts. Multiple versions, multiple venv solutions around.

To support all developers, it is far safer to just instruct PIP we require a venv. Python will handle it at that point.

Removing all venv logic and adding `export PIP_REQUIRE_VIRTUALENV=true` to the top of the file, protects the use without a venv, while keeping the freedom to use any tool or structure.

Although, the above, is my personal opinion.

balloob commented:
We should decide on what we want to achieve with the setup script and what it's responsibilities are.

From our `script/setup` [inspiration](https://github.com/github/scripts-to-rule-them-all#scriptsetup):

> script/setup is used to set up a project in an initial state. This is typically run after an initial clone, or, to reset the project back to its initial state.
> 
> This is also useful for ensuring that your bootstrapping actually works well.

The idea of putting the creation of the virtual environment into `script/setup` is because a virtual environment is part of the project's development environment and so part of `script/setup` to initialize.

What would be a use case for not using the built-in virtual environment of Python or bring your own venv?

pnbruckner committed:
Update setup

Check if venv active first. When looking for venv allow either `venv` or `.venv` for directory name, since both are common. Adjust how Python version is tested.

MartinHjelmare commented:
Personally I think the user interface of the standard venv isn't good. So I use virtualenv with virtualenvwrapper. That's one use case.

pnbruckner committed:
Specify valid Python versions in one place

bachya commented:
Silly example, but one that affects my workflow: I like to keep my venvs in `./.venv` (I even keep a series of Bash aliases that depend on that location). `script/setup` currently creates `./venv`, which kills my aliases. So, I end up manually creating `./.venv`, manually sourcing it, and manually running the appropriate pieces of `script/setup`. So, that's one case where `script/setup`'s opinionated creation of a venv hurts me.

pnbruckner committed:
Use HA_VENV_DIR if present

Allow user to specify where venv should be located.

pnbruckner commented:
So still not necessarily advocating either way, but to address previous comment I added the option for the user to specify where venv should be located via an environment variable, namely `HA_VENV_DIR`.

pnbruckner commented:
No consensus, so I'll just close this.
